### PR TITLE
update react-scripts from 4 to 5

### PR DESCRIPTION
**Pull request by benschac**
## What does this PR do and why?

**What**:
This PR takes the first step of updating the project's `react-scripts`. This is the output of running `yarn add --exact react-scripts@5.0.1`. I found this command in the changelog: https://github.com/facebook/create-react-app/blob/main/CHANGELOG.md

Notes: 

- this diff shouldn't be landed on its own but with the entire stack. Just running the single command above will leave the code base in a broken state.
- The lock file change is so big because CRA react-scripts@5 uses webpack@5 under the hood which drops all polyfills that the project used. 
- I'll continue to update my PRs as they're marked ready for review. There was significant concern about using re-wired in #712, which I'll address in that diff/PR.

**Why:**
#712 because we need to upgrade the project.

**How:**
Stacked Diffs: You'll see in the autogenerated comment below this that the following set of PRs are 'diffs' that are stacked on top of one another. Instead of each PR depending on `main`, they depend on each other, one on top of the other in sequence. 

**Why Stacked Diffs:**
Each diff is a single action in isolation, which makes each PR focused on a single problem or action. It's a pretty well-established way of working. Engineers at FB, Google, Airbnb and other big tech companies have been doing it for years. In my current role, we use Phabricator (https://www.phacility.com) which is a bit dated. Graphite (http://graphite.dev which is used to create this stacked diff is the modern version of phabricator that's built for GitHub.

It's a pretty common feature request of Github https://twitter.com/dan_abramov/status/1519020234316955650?s=20&t=wgrt4HlYmbfyKSGe1VkQfw (a common request of @gaearon too)

Here's a bit more documentation on stacked diffs if you're interested: 
- https://docs.graphite.dev/getting-started/why-use-stacked-changes
- https://graphite.dev/blog/post/DThX8ffP1gmxWJChEv0y
- https://graphite.dev/blog/post/N8SOs49y4bYdV1Vjf5qE

 

## Screenshots or screen recordings
N/A

## Acceptance checklist

- [x] I have evaluated the [Approval Guidelines](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#approval-guidelines) for this PR.
- [x] I have tested this PR in [all supported browsers](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#supported-browsers).

### CODE REVIEW

**Comment by benschac** [comment]
Current dependencies on/for this PR:
* main
  * **PR #867** <a href="https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/867" target="_blank">Graphite</a>  (this PR)
    * **PR #868** <a href="https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/868" target="_blank">Graphite</a> 
      * **PR #869** <a href="https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/869" target="_blank">Graphite</a> 
        * **PR #870** <a href="https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/870" target="_blank">Graphite</a> 
          * **PR #871** <a href="https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/871" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/867?utm_source=stack-comment).

**Comment by aeolianeth** [comment]
Big advocate of stacked diffs/chained PRs. Thanks @benschac, will review!

**Comment by aeolianeth** [comment]
@benschac heads up - I'm not going to get to this until next week. I've had a quick skim though and it looks really promising. Thanks a ton for diving into this 🙏

**Comment by benschac** [comment]
No worries, I wasn't planning on updating this stack until the weekend. Which, I should have communicated more directly here.


Ty for the heads up.

**Comment by aeolianeth** [comment]
@benschac can't thank you enough for your work here.

After reviewing these PRs and familiarizing myself with the CRA 5/polyfill status quo, it feels like moving forward with Rewire isn't as big of a commitment as I first thought. Correct me if I'm wrong, but it looks like, if CRA ever does decide to re-add support for those core node modules, we can quite easily strip out rewire and start using `react-scripts` to build the project again.

By committing to this though, we are still diverting from the out-of-the-box solution, which will naturally incur some maintenance overhead going forward (with more deps to keep track of and such). We're also becoming dependent on an "unofficial" tool, which despite being very well maintained, might prove to cause headaches down the road.

My only question at this point is: **what are we gaining from webpack 5/ CRA 5?** Are the benefits meaningful enough to warrant the above? I'm okay with staying with CRA 4 for the foreseeable future unless there is a compelling reason to upgrade. But if there are compelling reasons to upgrade, your proposed approach in these PRs is certainly the approach we should take 👍  LMK

**Comment by benschac** [comment]
> After reviewing these PRs and familiarizing myself with the CRA 5/polyfill status quo, it feels like moving forward with Rewire isn't as big of a commitment as I first thought. Correct me if I'm wrong, but it looks like, if CRA ever does decide to re-add support for those core node modules, we can quite easily strip out rewire and start using `react-scripts` to build the project again.
> 

Yes. Additionally, CRA needs a way to import polyfills now that webpack@5 no longer ships them by default. So, we'd have a similar setup to what we have done in this diff stack either way.


> My only question at this point is: **what are we gaining from webpack 5/ CRA 5?** Are the benefits meaningful enough to warrant the above? I'm okay with staying with CRA 4 for the foreseeable future unless there is a compelling reason to upgrade. But if there are compelling reasons to upgrade, your proposed approach in these PRs is certainly the approach we should take 👍 LMK

Yes, looking at the changelog I think it's worth keeping up to date with these deps and getting perf gains. I don't just think it's a `react-scripts` thing though. We'd get bit by the same issue, if there's a package we needed to use in `react-scripts@4` that needed a webpack configuration.

https://github.com/facebook/create-react-app/blob/main/CHANGELOG.md#full-changelog


**Webpack@5 is the biggest change: https://webpack.js.org/blog/2020-10-10-webpack-5-release/**
- Improve build performance with Persistent Caching.
- Improve Long Term Caching with better algorithms and defaults.
- Improve bundle size with better Tree Shaking and Code Generation.
- Improve compatibility with the web platform.
- Clean up internal structures that were left in a weird state while implementing features in v4 without introducing any breaking changes.
- Prepare for future features by introducing breaking changes now, allowing us to stay on v5 for as long as possible.

Debugging: 
- https://github.com/facebook/create-react-app/pull/8227


These are just my two cents. I'm not married to this solution and would be just fine abandoning this stack if the team feels like waiting is a better solution too. webpack@5 has been out for 2 years now and is pretty battle tested. re-wired looks like it has a lot of community support and large projects like web3.js are instructing users to use it directly in their documentation.

I'm not saying this implementation won't cause issues down that road, just that other teams and projects suggest this path and it seems fine.

**Comment by aeolianeth** [comment]
@benschac thanks for the detailed response. I'm on board. LFG.

How would you like to proceed? Shall we merge the PR chain from top to bottom into a single PR, test this PR, and merge it to main?

**Comment by aeolianeth** [comment]
cc @peripheralist @johnnyd-eth in case there are any concerns you have

**Comment by benschac** [comment]
> @benschac thanks for the detailed response. I'm on board. LFG.
> 
> How would you like to proceed? Shall we merge the PR chain from top to bottom into a single PR, test this PR, and merge it to main?

Yeah, I can merge them all together. Graphite has a merge feature but we'd need to add the graphite app to the repo. I'm not sure if folks want to do that.

File: src/components/Projects/index.tsx
Comment (benschac) [inline src/components/Projects/index.tsx:234]: Just trying to get `eslint`, curious how this got through CI.

File: src/declarations.d.ts
Comment (benschac) [inline src/declarations.d.ts:2]: https://bobbyhadz.com/blog/typescript-property-does-not-exist-on-type-window#fix---property-does-not-exist-on-type-window-in-typescript the article I found on using third party modules.

File: src/components/Projects/index.tsx
Comment (benschac) [reply src/components/Projects/index.tsx:234]: Wasn't sure if this is something I should get to the root of, or just get TSC to not complain.

File: src/components/v1/V1Project/BalanceTimeline.tsx
Comment (benschac) [inline src/components/v1/V1Project/BalanceTimeline.tsx:96]: @peripheralist -- Not sure how this was working. Assuming there's some kind of mismatch in the types or something?


I downloaded `"@types/ethereum-block-by-date": "^1.4.1",` types, but now that I think of it I'm really curious if there was a reason why they weren't initially added.


Just wanted to call this to your attn.

**Comment by benschac** [comment]
I still need to fix the tests. Please hold off on reviewing this PR for the moment.

**Comment by benschac** [comment]
Could I get access to fleek? I thought the easiest / and most logical path forward to fixing the failing CI tests would be to upgrade jest. Which, fixed the tests, but broke fleek's build and CI. I'm getting a 404 when clicking any of the fleek links.

CleanShot 2022-05-15 at 20 23 38@2x

**Comment by aeolianeth** [comment]
> Could I get access to fleek? I thought the easiest / and most logical path forward to fixing the failing CI tests would be to upgrade jest. Which, fixed the tests, but broke fleek's build and CI. I'm getting a 404 when clicking any of the fleek links.
> 
> CleanShot 2022-05-15 at 20 23 38@2x

@benschac ah gotcha. Unfortunately I think we've filled all of our Fleek seats, potentially @peripheralist can help check this.

In any case, here's the deploy log:

```
Deploy started at 4:33:50 AM 05/16/2022

4:33:51 AM 05/16/2022: Deploy started
4:33:53 AM 05/16/2022: Docker image: 'fleek/create-react-app'
4:33:53 AM 05/16/2022: Build command: 'yarn && yarn build'
4:33:53 AM 05/16/2022: Triggering build execution...
4:34:49 AM 05/16/2022: yarn install v1.22.5
4:34:50 AM 05/16/2022: [1/4] Resolving packages...
4:34:51 AM 05/16/2022: [2/4] Fetching packages...
4:36:00 AM 05/16/2022: info fsevents@2.3.2: The platform "linux" is incompatible with this module.
4:36:00 AM 05/16/2022: info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.
4:36:00 AM 05/16/2022: info fsevents@1.2.13: The platform "linux" is incompatible with this module.
4:36:00 AM 05/16/2022: info "fsevents@1.2.13" is an optional dependency and failed compatibility check. Excluding it from installation.
4:36:00 AM 05/16/2022: error jest-regex-util@28.0.0: The engine "node" is incompatible with this module. Expected version "^12.13.0 || ^14.15.0 || ^16.13.0 || >=17.0.0". Got "16.4.2"
4:36:01 AM 05/16/2022: error Found incompatible module.
4:36:01 AM 05/16/2022: info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
4:36:25 AM 05/16/2022: Error occured during the build.
```

**Review by wraeth-eth — CHANGES_REQUESTED** [review]
Looking good @benschac ! Just a couple of comments

File: src/components/shared/formItems/ProjectReserved.tsx
Comment (benschac) [reply src/components/shared/formItems/ProjectReserved.tsx:98]: Nah, this was a great catch. I shouldn't have changed this condition.

File: .eslintrc.json
Comment (benschac) [reply .eslintrc.json:36]: Ty, for commenting and giving me some direction. This was the initial stack these changes were included. I should have been a little more direct about asking for feedback. https://github.com/jbx-protocol/juice-interface/pull/876#issue-1216741357

I wasn't sure what the team wanted, and moving forward didn't want to change the code base too much, so I figured to do what was most sensible without introducing too many new rules right off the bat. I agree with most of your comments. I just wanted to get the ✅ before making all of these changes.

> One comment to cover some rules here:
> 
> * `@typescript-eslint/no-var-requires` - Do we want to allow requires? Isn't it better to use ES2015-style imports and use a disable if needed?

Yes, let's turn this one on.


> * `@typescript-eslint/ban-type` - I think we want to keep the [best practices](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/ban-types.md#default-options) here, wdyt?

👍🏻 -- here for it.

> * `react/no-children-prop` - i am not as well versed in react as others, but it is sometimes confusing and easier to miss children when passed in as props, I think it would be better to leverage the JSX and pass them in as DOM children.
👍🏻 SAME


> * `react/no-unescaped-entities` - This I think is a good one to keep to avoid accidentally leaking details into code.
👍🏻

> * `react/prop-types` - I would prefer to keep this one. I think it is less an issue with typescript as it likely already enforces it.
👍🏻

> * `prefer-const` - This one I think we should keep on too. Is there any benefit to not preferring const?
👍🏻

> * `jsx-a11y/img-redundant-alt` - seems like this is a good one to keep on for accessibility enforcement
👍🏻
> * `react/react-in-jsx-scope` - we definitely want this one I think to ensure that react is in scope

This one I don't think we need anymore: https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/react-in-jsx-scope.md#when-not-to-use-it it would be dope to remove some boilerplate. But I'm not married to the idea either.

**Comment by benschac** [comment]
So, I'm at a stopping point with this one. I've got tests and linting working both locally and on CI. Fleek continues to be a black-box, I don't really have any way of knowing what's going on there without seeing how the project is configured.


Next steps:
- Is there a way for me to get access to fleek?
- Is it possible for someone on the core team to take a look and debug it @aeolianeth @peripheralist @johnnyd-eth @wraeth-eth 
- Is it possible for me to get the exact configuration of fleek? I can fork the repo with my branch, debug on my end and come back with changes that need to be made on fleek
- Could I pair with someone on core, and we can take a look together?


I don't think it should be too time-consuming but I don't think I have enough information to complete this tasks ATM.

File: .eslintrc.json
Comment (wraeth-eth) [reply .eslintrc.json:36]: Oh awesome! I didn't actually realise that these violations already existed in the code. I am more leaning toward just pulling the trigger and fixing them in the PR - wdyt @aeolianeth ?

**Comment by wraeth-eth** [comment]
@aeolianeth is there anything we can do to help unblock @benschac here? I am a bit out of loop with the status of the fleek stuff, but is it possible (if we can't get more permissions) to get someone to have a look at fleek and see why this is failing?

**Comment by aeolianeth** [comment]
> @aeolianeth is there anything we can do to help unblock @benschac here? I am a bit out of loop with the status of the fleek stuff, but is it possible (if we can't get more permissions) to get someone to have a look at fleek and see why this is failing?

@wraeth-eth we've given @benschac Fleek access and he'll be tackling it this weekend 💪 https://discord.com/channels/939317843059679252/939705688563810304/978992950522572830

File: .eslintrc.json
Comment (aeolianeth) [reply .eslintrc.json:36]: @benschac @wraeth-eth i think we should resolve them but I don't want to block this PR any longer- happy for these changes to come in a follow-up

**Comment by aeolianeth** [comment]
Closing in favor of https://github.com/jbx-protocol/juice-interface/pull/1135

**Comment by aeolianeth** [comment]
@benschac thanks so so much for your hard work here.

Because we're moving to Next soon, I think we can stay on react-scripts@4 for now.

I've picked everything except for the rewired stuff into https://github.com/jbx-protocol/juice-interface/pull/1135

Closing this PR in favor of https://github.com/jbx-protocol/juice-interface/pull/1135

**Pull request closed without merging**
